File: src/main/java/com/kevinjanvier/resources/PetController.java
package com.kevinjanvier.resources;

import com.kevinjanvier.entity.Pet;
import com.kevinjanvier.service.PetService;
import io.micronaut.http.annotation.*;
import io.reactivex.Single;

import java.util.List;

@Controller("/api")
public class PetController {
    private final PetService petService;

    public PetController(PetService petService) {
        this.petService = petService;
    }

    @Post("/add-pet")
    public Pet savePet(@Body Pet pet){
        return petService.savePet(pet);
    }

    @Get("/get-all-pet")
    public Single<List<Pet>> getAllPet(){
        return Single.just(petService.getAllPets());
    }

    @Put("/update-pet")
    public Pet updatePet(@Body Pet pet){
        return petService.updatePet(pet);
    }

    @Delete("/delete-pet")
    public void deletePet(@PathVariable("id")Long id){
        petService.deletePet(id);
    }
}

File: src/main/kotlin/com/kevinjanvier/resources/PetController.kt
package com.kevinjanvier.resources

import com.kevinjanvier.service.PetService
import com.kevinjanvier.entity.PetModel
import com.kevinjanvier.response.PetResponse
import io.micronaut.http.HttpResponse
import io.micronaut.http.annotation.*
import java.net.URI

@Controller("/api")
class PetController(private val petService: PetService) {


    @Post("/create")
    fun create(@Body request: PetModel): HttpResponse<Void> {
        val createdId = petService.createPet(request)

        return HttpResponse.created(
                URI.create(
                        createdId!!.asObjectId().value.toHexString(
                        )
                ))
    }

    @Get
    fun findAll(): HttpResponse<List<PetResponse>> {
        val pets = petService
                .findAll()
                .map { PetResponse.fromEntity(it) }

        return HttpResponse.ok(pets)
    }

    @Get("/{id}")
    fun findById(@PathVariable id: String): HttpResponse<PetResponse> {
        val pets = petService.findById(id)
        return HttpResponse.ok(
                PetResponse.fromEntity(pets)
        )
    }

    @Put("/update-pet/{id}")
    fun updatePet(
            @PathVariable id: String,
            @Body petModel: PetModel?): HttpResponse<PetResponse> {

        var updatedPet = petService.updatePet(id, petModel!!)
        return HttpResponse.ok(
                PetResponse.fromEntity(updatedPet)
        )

    }

    @Delete("/delete-pet/{id}")
    fun deletePetById(@PathVariable("id") id: String?): HttpResponse<Void> {
        petService.deleteById(id!!)
        return HttpResponse.noContent()
    }
}
